### PR TITLE
Rename DeploymentConfig to Props

### PR DESCRIPTION
**Pull request by johanandren**
Refs #22660

### CODE REVIEW

File: akka-typed/src/main/scala/akka/typed/Deployment.scala [hunk 1]
```diff
@@ -18,97 +21,108 @@ import scala.annotation.tailrec
...
- * The DeploymentConfig includes a `next` reference so that it can form an
+ * The Props includes a `next` reference so that it can form an
  * internally linked list. Traversal of this list stops when encountering the
- * [[EmptyDeploymentConfig$]] object.
+ * [[EmptyProps]] object.
  */
-abstract class DeploymentConfig extends Product with Serializable {
+@DoNotInherit
+@ApiMayChange
+abstract class Props private[akka] () extends Product with Serializable {
```
Comment (patriknw) [inline akka-typed/src/main/scala/akka/typed/Deployment.scala]: rename the file to Props.scala

File: akka-typed/src/main/scala/akka/typed/Deployment.scala [hunk 2]
```diff
@@ -120,17 +134,20 @@ abstract class DeploymentConfig extends Product with Serializable {
...
 
 /**
  * The empty configuration node, used as a terminator for the internally linked
- * list of each DeploymentConfig.
+ * list of each Props.
  */
-case object EmptyDeploymentConfig extends DeploymentConfig {
-  override def next = throw new NoSuchElementException("EmptyDeploymentConfig has no next")
-  override def withNext(next: DeploymentConfig): DeploymentConfig = next
+case object EmptyProps extends Props {
```
Comment (patriknw) [inline akka-typed/src/main/scala/akka/typed/Deployment.scala]: I think this should be internal, and user api should be `Props.empty`

**Comment by johanandren** [comment]
Was a bit quick opening the PR there, will push an update shortly.

**Review by patriknw — CHANGES_REQUESTED** [review]
looking good, small adjustment...

File: akka-typed/src/main/scala/akka/typed/Deployment.scala [hunk 3]
```diff
@@ -134,20 +148,19 @@ abstract class Props private[akka] () extends Product with Serializable {
  * The default mailbox capacity that is used when this option is not given is
  * taken from the `akka.typed.mailbox-capacity` configuration setting.
```
Comment (ktoso) [inline akka-typed/src/main/scala/akka/typed/Deployment.scala]: also INTERNAL API in the comment please

File: akka-typed/src/main/scala/akka/typed/Deployment.scala [hunk 4]
```diff
@@ -134,20 +148,19 @@ abstract class Props private[akka] () extends Product with Serializable {
...
-  @InternalApi
+@InternalApi
+private[akka] final case class MailboxCapacity(capacity: Int, next: Props = Props.empty) extends Props {
   private[akka] override def withNext(next: Props): Props = copy(next = next)
 }
 
 /**
  * The empty configuration node, used as a terminator for the internally linked
  * list of each Props.
  */
```
Comment (ktoso) [inline akka-typed/src/main/scala/akka/typed/Deployment.scala]: INTERNAL API in comment

File: akka-typed/src/main/scala/akka/typed/Deployment.scala [hunk 5]
```diff
@@ -156,17 +169,23 @@ case object EmptyProps extends Props {
  *
  * The default configuration if none of these options are present is to run
  * the actor on the same executor as its parent.
+ *
+ * INTERNAL API
  */
-sealed trait DispatcherSelector extends Props
+@DoNotInherit
```
Comment (ktoso) [inline akka-typed/src/main/scala/akka/typed/Deployment.scala]: Internal implies "do not touch at all, also dont sub class" I'd just keep that one

**Review by ktoso — APPROVED** [review]
LGTM after file rename

File: akka-typed/src/main/scala/akka/typed/Props.scala [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+
+  /**
+   * Create a copy of this Props node with its `next` reference
+   * replaced by the given object. <b>This does NOT append the given list
+   * of configuration nodes to the current list!</b>
+   *
+   * INTERNAL API
+   */
+  @InternalApi
+  private[akka] def withNext(next: Props): Props
```
Comment (patriknw) [inline akka-typed/src/main/scala/akka/typed/Props.scala]: now it's not "open" any more, perhaps keep it but name it `with(props: Props)` and document if it is used before or after other props of same type

Comment (ktoso) [inline akka-typed/src/main/scala/akka/typed/Props.scala]: `with` will conflict with keyword :(

Comment (johanandren) [inline akka-typed/src/main/scala/akka/typed/Props.scala]: Should it be open, I was under the impression it is only about being for our own implementations (cluster for example)?

Comment (patriknw) [inline akka-typed/src/main/scala/akka/typed/Props.scala]: ah, `private[akka]` is enough for that, sorry for the noise

**Review by patriknw — APPROVED** [review]
LGTM

**Pull request merged**